pennylane/tape/tape.py

```diff
@@ -1555,8 +1589,18 @@ def copy(self, copy_operations=False):
             tape._ops = self._ops.copy()
             tape._measurements = self._measurements.copy()
 
-        tape._update()
+        tape._graph = self._graph
+        tape._specs = self._specs
+        tape._depth = self._depth
```

Comment:
Reverting this back to 
```suggestion
        tape._graph = None
        tape._specs = None
        tape._depth = None
```
because these attributes need to be constructed newly for the new tape. This is found by failing tests, I did not dig through the entire logic of `_graphs` or `_specs`

Reply:
Mhhh. But if the tape is a copy, then shouldn't these be the same? Maybe this is also a shallow/deep copy issue?

Reply:
The issue is with `copy_operations`: If it is set to `True`, the operations in the new tape will be new `Operation` instances.
Therefore, when copying over the old graph, the new instances will not be found in that graph, making lookups etc fail.

We could use copy over the `_graph` in the case `copy_operations=False` (the default) though :) But for example the parameter shift rule makes use of `copy_operations=True`, and any other application that wants to alter the operation parameters later on will do so, too...

Reply:
Does this mean that each tape holds references to its Operations in multiple independent data structures and needs to do lookups between them? Naively this sounds sub-optimal. The graph is probably the most "complete" and "elaborate" description of the circuit. Couldn't things like `tape.operations` draw from the graph? Maybe this is beyond this PR... :-)

Reply:
Yes, I think there might be some redundancy. I'm not sure how efficient it would be to always draw the information from the graph, in particular, because `_graph` is not computed at initialization of tapes, but only when calling the `graph` property.
This means one can use `QuantumTape`s more or less as a list, and causal dependencies in the circuit are only considered once we want to actually do that :)